rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,716 @@
+   potentially useful information.
+
+4. We could grant bots permission to view their own join tokens. There is
+   precedent here as bots can view e.g. their own roles without explicitly
+   having RBAC permissions to do so.
+
+The remaining rejoin counter should then be exposed as a Prometheus metric to
+allow for alerting if a bot drops below some threshold.
+
+Importantly, this is a potentially lagging indicator. The design allows for the
+rejoin counter to be decreased (to zero) at any time, so a rejoin attempt may
+still fail at any time. This should be acceptable since it can also be increased
+after the fact to restore access if desired.
+
+#### Keystore Storage Backends
+
+We should support abstract keystore storage backends to enable storage methods
+beyond plain file storage.
+
+For example:
+
+- HSM storage, for hardware supporting PKCS#11. This includes many TPM 1.2 / 2.0
+  implementations. We have some prior art here in Teleport's [HSM
+  support](https://goteleport.com/docs/admin-guides/deploy-a-cluster/hsm/).
+
+- [Apple Secure Enclave key storage][enclave]. This would require additional
+  changes to our release process, as access to this functionality requires app
+  signing. We again have prior art here with `tsh`.
+
+Further evaluation will be necessary to ensure these backends support our
+challenge process and key types. Libraries like [`sks`] provide compatibility
+across TPM 2.0 (Windows, Linux) and Apple's Secure Enclave, and should be able
+to sign our challenges appropriately, and using our desired key types.
+
+[enclave]: https://developer.apple.com/documentation/security/protecting-keys-with-the-secure-enclave
+[`sks`]: https://github.com/facebookincubator/sks
+
+#### Non-Terraform UX
+
+This proposal also aims to improve the non-Terraform UX, particularly when
+automating with `tctl`. All regular token management workflows with
+`tctl create -f` will continue to work; upserting resources to modify runtime
+values will, for example, properly increase
+`status.bound_keypair.remaining_rejoins` while preserving other token fields
+like `status.bound_keypair.bound_public_key`.
+
+Additional `tctl` changes will include:
+
+- Once satisfied with behavior of the join method, replacing the default
+  automatically generated join tokens for `tctl bots add` and
+  `tctl bots instances add` to use this new join method.
+
+- Adding a column for "rejoins remaining" in `tctl bots instances ls` (where
+  relevant).
+
+- Adding support for updating `total_rejoins` in `tctl bots update`
+
+#### Expiration Alerting UX
+
+A major deficiency in `token` joining is that bots fail silently. Their status
+can be partially monitored via `tctl get bot_instance` but this does not
+effectively notify administrators when something has gone wrong.
+
+We should take steps to improve visibility of bots at or near expiry, including:
+
+- Configurable cluster alerts when the number of available renewals has crossed
```

Comment:
Maybe add an alert that states something like: "A bot has recently automatically rejoined", without specifically watching the count.
Where 'recently' is configurable.

Reply:
That's a good thought, I've added a note about it to this section. There's some mild complexity here, I think, since alerts could get noisy if there's hundreds (or more) bots active. I think we'll definitely need to iterate on this as we see how this method gets used.